### PR TITLE
Hyperchains parent manager

### PR DESCRIPTION
**Pull request by erlmachinedev**
This PR contains the next described functionality + test SUITE which supplied with a simple parent chain demo generator:

### Hyperchains parent's process manager
See the pattern "ProcessManager (https://www.enterpriseintegrationpatterns.com/patterns/messaging/ProcessManager.html).

This component is responsible for orchestration the abstract parent's chain layer through:
 a) dedicated workers (trackers);
 b) supplied interface providers (connectors);

### Notes
 - The first supplied connector acts as the "master" and is eligible to dictate the election period through the event's announcement;
 - A tracker performs synchronization from the highest known block until the genesis pointer through the "previous_hash" property;
 - To be able to run the Hyperchains the system must satisfy the connector's acceptance criteria;
     a) default mode: get_top_block/0, get_block_by_hash/1l
     b) delegate mode: default mode + send_tx/1;
 - Each interested developer can supply their own connector's implementation over the particular parent chain;

### CODE REVIEW

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (radrow) [inline apps/aehyperchains/src/aehc_parent_mng.erl:63]:
```suggestion
    {ok, pid()} | ignore | {error, term()}.
```

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (radrow) [inline apps/aehyperchains/src/aehc_parent_mng.erl:149]:
```suggestion
        [Criteria(Track) || Criteria <- [fun accept_top_block/1, fun accept_block_by_hash/1, fun accept_send_tx/1]],
```

**Review by radrow — CHANGES_REQUESTED** [review]

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:180]: You don't submit the commitment hash

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:181]: I don't like this sanity check

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:206]: This is blocking...

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:91]: This is blocking... This needs to be done in separate processes

**Review by gorbak25 — CHANGES_REQUESTED** [review]
To summarize:
1. More tests, perhaps develop more in line with TDD.
2. Use monitors... instead of trapping exits...
3. Tracks should represent "A sync job which is currently under way"
4. Each track should be processed in it's own process - coordinated by the connector manager which provides an API for the currently active jobs
5. The master connector should not block for a week in the init function - just generate some initial jobs,  monitor the job pids and enter the genserver loop...
6. Sync tasks should be parallelized - The manager should keep track of the amount of jobs for a given connector, If a given sync job(track) can be split in half the do it if we have a small ammout of workers currently active.

7. Besides that good job! Unfortunately this PR still requires a lot of work.

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:6]: Thanks

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:7]: Thanks

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:9]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:10]: Thanks

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:11]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:12]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:18]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:63]: +

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:149]: Oh, thanks

File: apps/aehyperchains/src/aehc_connector.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:29]: I faced some issue, have to disscuss. Our API allows to create by:

`-spec new(aec_headers:key_header(), aec_headers:key_header()) -> pogf().`

But we have to be able to reproduce PoGF object from payload record as is, I mean something like:

`-spec new(binary()), binary())) -> pogf().`

Moreover, aec_headers:key_header() entries duplicate data from the current existing ones (already persisted on blockchain).

I propouse to reduce type pogf() into hash storing document

File: apps/aehyperchains/src/aehc_parent_block.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_block.erl:91]: Sorry, can't see what is wrong here. This function acts as a filter within manager subsribtion duties, to check arrived message for correctness from a hight level type perspective.

File: apps/aehyperchains/src/aehc_app.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_app.erl:25]: Thanks, also think that hyperchains section should be placed at higher level

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:181]: As you will wish, sir

File: apps/aehyperchains/src/aehc_commitment.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_commitment.erl:77]: I can explain the need. The connector module should be able to operate payload data (at least to provide signature check we have to extract **"delegate"** piece of data). The intention is to provide payload accessor which can take binary payload as an argument and to extract particular data from it.

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:83]: Oh, you know I appreciate your intention to improve (that's great quality). 
But, please note, that after we wrapped Db API into transaction calls - that all doesn't matter. I hope you know how transaction mechanism works, we work around database and eventially our DB (and we are too) is sync.

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:90]: Agree, manager should operate independent processes, sure

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_mng.erl:208]: Yep, you are right, but:

1) Are we eligible to proceed until one of the connectors isn't ready? Sounds simple, but actually I am not sure about;
2) I mention you transaction consequences below. We work around database and this table each time in a "freeze" state until call has executed. That means that a lot of parallelizing features don't impact here.

So, in general I agree that we have to decouple job into processes, and that's clear for me. But please also note the sync nature of some init procedures which we can't skip.

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (gorbak25) [reply apps/aehyperchains/src/aehc_parent_mng.erl]: Your argument is valid ONLY when the time to retrieve data from the parent chain is negligible which might be the case for our demo "chain simulator" connector.
In the real world the time to retrieve blocks and commitments from the parent chain is significant and can't be ignored. Spiting up tracks will yield an increase in performance for:
- Connectors using HTTP to retrieve data from an API endpoint
- Connectors which are essentially light clients and validate data provided by other peers

Anyway - if we got an async job infrastructure setup then why not parallelize the jobs?

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (gorbak25) [reply apps/aehyperchains/src/aehc_parent_mng.erl]: Other parts of the node might want to inspect the status of the parent sync - especially a user wondering why the node isn't fully booting. Without entering the genserver loop quickly we essentially refuse to respond to requests about the status of the sync jobs. Please take a look at aec_sync which has a similar problem.

After sync is done the mining conductor should get notified that "Hey we are ready to participate in elections".

File: apps/aehyperchains/src/aehc_connector.erl
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_connector.erl:44]: Ok, this helper is still broken and generates an invalid parent block ADT.
Please preserve the following invariant:
```
is_list(Block.Header.commitment_hashes),
is_list(Block.commitments),
length(Block.Header.commitment_hashes) == length(Block.commitments),
[Hash = aehc_commitment:hash(Commitment) || {Hash, Commitment} <- lists:zip(Block.Header.commitment_hashes, Block.commitments)]
```
This is essentially the same invariant we preserve when handling microblocks. 
Take a look at:
https://github.com/aeternity/aeternity/blob/8b33ad36d6f81d90f57178b735819c01bfe652c8/apps/aehyperchains/src/aehc_parent_db.erl#L96

File: apps/aehyperchains/src/aehc_parent_db.erl
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_db.erl:162]:
```suggestion
find_parent_chain_view(GenesisHash) ->
```

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:14]: send_tx shouldn't send out a real tx when initializing the system

File: apps/aehyperchains/src/aehc_parent_mng.erl
Comment (gorbak25) [inline apps/aehyperchains/src/aehc_parent_mng.erl:85]:
```suggestion
    %% To be able to persists the block we have to be sure that it connects to the previous existing one in the DB;
```

File: apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl
Comment (gorbak25) [inline apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl:65]: You are sending the commitment ID instead of the contents.

**Review by gorbak25 — CHANGES_REQUESTED** [review]
I like the direction this PR is moving but some small work still needs to be done before it can be merged:
- Only empty blocks are properly saved in the DB - the code fails for blocks with commitments
- aehc_parent_mng:start_view should be called by aehc_parent_mng and not the user
- Improper usage of commitment hashes - a commitment hash is a unique identifier not the contents we send out to the parent chain

File: apps/aehyperchains/src/aehc_connector.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:44]: How much beer have you drunk yesterday that posted Elixir syntax here? Joke :) Missed one argument in constructor, thanks for the notice.

File: apps/aehyperchains/src/aehc_connector.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:46]: **1, 3** - got you, will fix today
About **2**, just short notice to be sure that we at "the same ground here". The main design consideration around parent_mng is the ability to manage the interaction between multiple parent chains, like:

1) Who dictates the election period?
2) What we can do in a conflict situation that arose between multiple parent views?
3) Some cases can come later and this component can be improved by additional functionality;

So it's not a supervisor, for the monitor purposes we have - supervisor :). API is called by our app and is exported with keeping in mind the dynamic nature of HC - when something bad happens outside the operator should be able to switch or reject some chain.

File: apps/aehyperchains/src/aehc_connector.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:111]: Maybe you are right, just over measure from me. Idea with dry_send_tx seems ok

Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:111]: We perform now two checks: get_top_block(), get_block_by_hash(Genesis), and both of them are independent and have to be satisfied;

Implementation of dry_send_tx can be delegated to a particular connector:

dry_send_tx -> just to return **ok** (if I am not going to spend fee);
dry_send_tx -> more specific check, like contract existence, available balance etc.

File: apps/aehyperchains/src/aehc_connector.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:135]: Today you have that knowledge that "It's a one-time conversion", tomorrow me, or new Adam, or somebody else will use that in another way - and we got node leak.

These rules have written by blood - I am not going to joke here). I will mark this in code with your suggestion.

File: apps/aehyperchains/src/aehc_parent_db.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_db.erl:154]: ok

File: apps/aehyperchains/src/aehc_parent_db.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_db.erl:162]: write_, get_ are convenient prefixes here.

File: apps/aehyperchains/src/aehc_parent_db.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_db.erl:173]: This API reflects aec_chain:

top_block/0 -> get_parent_top_block/0;
get_block/1 -> get_parent_block/1

**Review by gorbak25 — CHANGES_REQUESTED** [review]
Good job! We are close to merging this PR.
2 small things remain before merging this to HC-master:
- The commits will be squashed anyway but please rebase your branch and ensure that the commits are more or less clear (69 commits with multiple merges of hyperchains-master + duplicated commits due to merging instead of rebasing is extremely unclear). It doesn't have to be perfect but at least remove the standing out things - merges, duplicates etc...
- Add a test case for saving a block with commitments via a connector and reading it back from DB - currently we only test blocks without commitments

Optional:
Enable GPG auto signing of your commits.

**Review by gorbak25 — APPROVED** [review]

File: apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl
Comment (radrow) [inline apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl:22]: why so unspecified?

File: apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl
Comment (radrow) [reply apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl:71]: +1

File: apps/aehyperchains/src/aehc_parent_block.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_block.erl:91]: +

File: apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl
Comment (erlmachinedev) [reply apps/aehyperchains/test/connectors/aehc_chain_sim_connector.erl:22]: +

File: apps/aehyperchains/src/aehc_parent_block.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_parent_block.erl:20]: prev_hash_block, hash_block, height_block - for the sake of compatibility

File: apps/aehyperchains/src/aehc_connector.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:103]: It's short notice, works for me (but I got you if you see some mess here I remove)

File: apps/aehyperchains/src/aehc_connector.erl
Comment (erlmachinedev) [reply apps/aehyperchains/src/aehc_connector.erl:35]: Yep, you are right - fixed

**Pull request merged**
